src/icrc1_auction_api_new.mo

```diff
@@ -128,12 +128,125 @@ actor class Icrc1AuctionAPI(quoteLedger_ : ?Principal, adminPrincipal_ : ?Princi
     };
   };
 
-  var auction : ?Auction.Auction = null;
-  var assets : Vec.Vector<AssetInfo> = Vec.new();
+  let quoteAssetId : Auction.AssetId = 0;
+
+  let assets : Vec.Vector<AssetInfo> = Vec.map<StableAssetInfo, AssetInfo>(
+    assetsDataV3,
+    func(x) {
+      let r = (actor (Principal.toText(x.ledgerPrincipal)) : ICRC1.ICRC1Ledger)
+      |> {
+        balance_of = _.icrc1_balance_of;
+        fee = _.icrc1_fee;
+        transfer = _.icrc1_transfer;
+        transfer_from = _.icrc2_transfer_from;
+      }
+      |> {
+        ledgerPrincipal = x.ledgerPrincipal;
+        minAskVolume = x.minAskVolume;
+        handler = TokenHandler.TokenHandler({
+          ledgerApi = _;
+          ownPrincipal = Principal.fromActor(self);
+          initialFee = 0;
+          triggerOnNotifications = true;
+          log = func(p : Principal, logEvent : TokenHandler.LogEvent) = Vec.add(tokenHandlersJournal, (x.ledgerPrincipal, p, logEvent));
+        });
+        decimals = x.decimals;
+        symbol = x.symbol;
+      };
+      r.handler.unshare(x.handler);
+      r;
+    },
+  );
+  let auction : Auction.Auction = Auction.Auction(
+    0,
+    {
+      volumeStepLog10 = 3; // minimum quote volume step 1_000
+      minVolumeSteps = 5; // minimum quote volume is 5_000
+      priceMaxDigits = 5;
+      minAskVolume = func(assetId, _) = Vec.get(assets, assetId).minAskVolume;
+      performanceCounter = Prim.performanceCounter;
+    },
+  );
+  auction.unshare(auctionDataV7);
+
+  // each 2 minutes
+  let AUCTION_INTERVAL_SECONDS : Nat64 = 120;
+  var nextSessionTimestamp = Nat64.toNat(AUCTION_INTERVAL_SECONDS * (1 + Prim.time() / (AUCTION_INTERVAL_SECONDS * 1_000_000_000)));
+
+  private func registerAsset_(ledgerPrincipal : Principal, minAskVolume : Nat) : async* R.Result<Nat, RegisterAssetError> {
+    let canister = actor (Principal.toText(ledgerPrincipal)) : (actor { icrc1_decimals : () -> async Nat8; icrc1_symbol : () -> async Text });
+    let decimals = try {
+      Nat8.toNat(await canister.icrc1_decimals());
+    } catch (err) {
+      throw err;
+    };
+    let symbol = try {
+      await canister.icrc1_symbol();
+    } catch (err) {
+      throw err;
+    };
+    for ((assetInfo, i) in Vec.items(assets)) {
+      if (Principal.equal(ledgerPrincipal, assetInfo.ledgerPrincipal)) return #err(#AlreadyRegistered(i));
+    };
```

Comment:
We should use Vec.contains here instead of the for loop. Or Vec.forSome

Reply:
Done